inst/tests/tests.Rraw

```diff
@@ -3729,6 +3729,15 @@ test(1100, dt1[dt2,roll=-Inf,rollends=c(FALSE,TRUE)]$ind, INT(NA,NA,1,2,2,2,2,2,
   DT = data.table(x=sample(5,20,TRUE), y=sample(2,20,TRUE), z=sample(letters[1:2],20,TRUE), d1=runif(20), d2=1L)
   test(1102.38, names(dcast(DT, x ~ y + z, fun.aggregate=length, value.var = "d2", sep=".")),
                 c("x", "1.a", "1.b", "2.a", "2.b"))
+
+  # test for #5512, only compute default fill if needed.
+  DT = data.table(chr=c("a","b","b"), int=1:3, dbl=as.double(4:6))
+  test(1102.39, dcast(DT, int ~ chr, min, value.var="int"), data.table(int=1:3, a=c(1L,NA,NA), b=c(NA,2L,3L), key="int"), warning="NAs introduced by coercion to integer range")#as.integer(min(integer()) is Inf) is NA
+  test(1102.391, dcast(DT, int ~ chr, min, value.var="int", fill=NA), data.table(int=1:3, a=c(1L,NA,NA), b=c(NA,2L,3L), key="int"))
+  test(1102.392, dcast(DT, int ~ chr, min, value.var="dbl"), data.table(int=1:3, a=c(4,Inf,Inf), b=c(Inf,5,6), key="int"))
+  test(1102.40, dcast(DT, . ~ chr, min, value.var="int"), data.table(.=".",a=1L,b=2L,key="."))
+  test(1102.41, dcast(data.table(a = 1, b = 2, c = 3), a ~ b, value.var = 'c', fill = '2'), data.table(a=1, `2`=3, key='a'))
```

Comment:
If I'm understanding this test correctly, I think it would be clearer to put this test first, it is closest to the spirit of the change described (don't calculate 'fill' if not needed). Then a quick comment transitioning to the tests about potential warnings emitted from `fun.aggregate=` in the edge case.

Reply:
Does this change work with `fill = stop("shouldn't evaluate")`? if so, that might make the test as clear as possible. if not, why not?

Reply:
not it does not, because standard evaluation is used (fill is always evaluated, even if not used)

Reply:
and I think this test is closest to the spirit of the change, so I put it first:
```r
test(1102.39, dcast(DT, . ~ chr, min, value.var="int"), data.table(.=".",a=1L,b=2L,key="."))#fill not used in output, so default fill not computed, and no warning emitted during computation of that default fill.
```

Reply:
`fill = stop("shouldn't evaluate")` could be implemented if we hack fcast.R (this PR only touches fcast.c)

Reply:
thanks for clarifying! i don't think there's strong reason to do that. By far the typical use case for `fill=` supplies a literal value; [I don't see any case where `fill = foo()`](https://github.com/search?q=lang%3AR+%2F%28%5B%5Ea-z0-9._%3A%5D%7Ce%3A%3A%29dcast%5B%28%5D%5B%5E%29%5D%2F+%2F%5B%5Ea-z0-9._%5Dfill%5Cs*%3D%5Cs*%5Ba-z0-9._%5D%2B%5B%28%5D%2F+NOT+%2Fgeom_%2F&type=code) where `foo()` might be slow & we'd benefit from lazy evaluation.

We can re-evaluate if there's a compelling FR later.

Reply:
great